LimeSuiteNG soapy stream not starting with certain applications

Lubuntu 24.04 with latest LimeSuiteNG using a Lime Mini v2 as of this morning. 
Soapy Lib Version: v0.8.1-4build1
API Version: v0.8.0
ABI Version: v0.8

I recall something very similar happening to a fork of both the Airspy and AirspyHF soapy plugins where in some applications due to whatever reason, the stream didn't start fast enough? 

Rechecking CubicSDR soapy results. It would appear the stream never starts.

```
[INFO] Make connection: 'LimeSDR Mini xxxxxxx'
[INFO] Reference clock 40.00 MHz
[INFO] Device name: LimeSDR-Mini_v2
[INFO] Reference: 40 MHz
[INFO] LMS7002M register cache: Disabled
SDR thread starting.
device init()
[INFO] Sampling rate set(2.100 MHz): CGEN:268.800 MHz, Decim: 2^5, Interp: 2^5
[INFO] Rx LPF configured
[ERROR] LoadConfig() - file not found
[INFO] Sampling rate set(2.100 MHz): CGEN:268.800 MHz, Decim: 2^5, Interp: 2^5
[INFO] Sampling rate set(2.100 MHz): CGEN:268.800 MHz, Decim: 2^5, Interp: 2^5
SDRThread : Device Stream set to MTU: 8160
starting readLoop()
```


Rechecking GQRX soapy results. Also appears the stream does not start.
```
[INFO] Make connection: 'LimeSDR Mini xxxxxx'
[INFO] Reference clock 40.00 MHz
[INFO] Device name: LimeSDR-Mini_v2
[INFO] Reference: 40 MHz
[INFO] LMS7002M register cache: Disabled
[INFO] Sampling rate set(10.003 MHz): CGEN:320.104 MHz, Decim: 2^3, Interp: 2^3
[INFO] Rx LPF configured
[INFO] Sampling rate set(5.000 MHz): CGEN:640.000 MHz, Decim: 2^5, Interp: 2^5
[INFO] Rx LPF configured
[ERROR] No samples or timeout
[ERROR] No samples or timeout
[ERROR] No samples or timeout
[ERROR] No samples or timeout
[ERROR] No samples or timeout
[ERROR] No samples or timeout
```

SigDigger. Stream starts and works. 
SDRAngel with Soapy. Stream starts and works.

--- a/plugins/soapysdr/Streaming.cpp
+++ b/plugins/soapysdr/Streaming.cpp
@@ -176,8 +176,18 @@ SoapySDR::Stream* Soapy_limesuiteng::setupStream(
     //     config.bufferLength = std::stoul(args.at("bufferLength"));
     // }
 
-    // Stream creation is deferred to activateStream(), the device may not be
-    // fully configured (e.g. sample rate) at this point
+    // Setup is deliberately split from start so that activateStream() stays fast.
+    // The sample rate only hints buffering sizes, so when the app has not
+    // configured it yet, fall back to the device's current rate.
+    if (sampleRate[SOAPY_SDR_RX] <= 0.0)
+        sampleRate[SOAPY_SDR_RX] = sdrDevice->GetSampleRate(0, TRXDir::Rx, 0);
+    config.hintSampleRate = sampleRate[SOAPY_SDR_RX];
+
+    rfstream = sdrDevice->StreamCreate(config, 0);
+    if (!rfstream)
+    {
+        throw std::runtime_error("Soapy_limesuiteng::setupStream() failed: " + std::string(GetLastErrorMessage()));
+    }
 
     auto samplesPerPacket = config.linkFormat == DataFormat::I16 ? 1020 : 1360;
     // TODO: figure out a way to actually get the correct packet per batch count
@@ -237,8 +247,11 @@ int Soapy_limesuiteng::activateStream(SoapySDR::Stream* stream, const int flags,
     //     settingsCache.at(direction).at(ch).calibrationBandwidth = bw;
     //     _channelsToCal.erase(_channelsToCal.begin());
     // }
-    if (!rfstream)
+    // recreate the stream only if the rate it was built with no longer matches,
+    // e.g. the app configured the sample rate after setting up the stream
+    if (!rfstream || streamConfig.hintSampleRate != sampleRate[SOAPY_SDR_RX])
     {
+        streamConfig.hintSampleRate = sampleRate[SOAPY_SDR_RX];
         rfstream = sdrDevice->StreamCreate(streamConfig, 0);
         if (!rfstream)
             throw std::runtime_error("Soapy_limesuiteng::activateStream() failed: " + std::string(GetLastErrorMessage()));

--- a/plugins/soapysdr/Streaming.cpp
+++ b/plugins/soapysdr/Streaming.cpp
@@ -176,13 +176,8 @@ SoapySDR::Stream* Soapy_limesuiteng::setupStream(
     //     config.bufferLength = std::stoul(args.at("bufferLength"));
     // }
 
-    // Create the stream
-
-    rfstream = sdrDevice->StreamCreate(config, 0);
-    if (!rfstream)
-    {
-        throw std::runtime_error("Soapy_limesuiteng::setupStream() failed: " + std::string(GetLastErrorMessage()));
-    }
+    // Stream creation is deferred to activateStream(), the device may not be
+    // fully configured (e.g. sample rate) at this point
 
     auto samplesPerPacket = config.linkFormat == DataFormat::I16 ? 1020 : 1360;
     // TODO: figure out a way to actually get the correct packet per batch count
@@ -198,8 +193,11 @@ void Soapy_limesuiteng::closeStream(SoapySDR::Stream* stream)
 {
     std::unique_lock<std::recursive_mutex> lock(_accessMutex);
     auto icstream = reinterpret_cast<IConnectionStream*>(stream);
-    rfstream->Stop();
-    rfstream.reset();
+    if (rfstream)
+    {
+        rfstream->Stop();
+        rfstream.reset();
+    }
     delete icstream;
 }
 
@@ -239,6 +237,13 @@ int Soapy_limesuiteng::activateStream(SoapySDR::Stream* stream, const int flags,
     //     settingsCache.at(direction).at(ch).calibrationBandwidth = bw;
     //     _channelsToCal.erase(_channelsToCal.begin());
     // }
+    if (!rfstream)
+    {
+        rfstream = sdrDevice->StreamCreate(streamConfig, 0);
+        if (!rfstream)
+            throw std::runtime_error("Soapy_limesuiteng::activateStream() failed: " + std::string(GetLastErrorMessage()));
+    }
+
     // Stream requests used with rx
     icstream->flags = flags;
     icstream->rxBurstStart_timeNs = timeNs;
@@ -256,7 +261,8 @@ int Soapy_limesuiteng::deactivateStream(
     std::unique_lock<std::recursive_mutex> lock(_accessMutex);
     auto icstream = reinterpret_cast<IConnectionStream*>(stream);
     icstream->rxBurstRequest = false;
-    rfstream->Stop();
+    if (rfstream)
+        rfstream->Stop();
     isStreamRunning = false;
     return 0;
 }
@@ -268,6 +274,8 @@ int Soapy_limesuiteng::readStream(
     SoapySDR::Stream* stream, void* const* buffs, size_t numElems, int& flags, long long& timeNs, const long timeoutUs)
 {
     auto icstream = reinterpret_cast<IConnectionStream*>(stream);
+    if (!rfstream)
+        return SOAPY_SDR_STREAM_ERROR;
 
     // Handle the one packet flag by clipping
     if ((flags & SOAPY_SDR_ONE_PACKET) != 0)
@@ -374,6 +382,8 @@ int Soapy_limesuiteng::writeStream(SoapySDR::Stream* stream,
     }
 
     auto icstream = reinterpret_cast<IConnectionStream*>(stream);
+    if (!rfstream)
+        return SOAPY_SDR_STREAM_ERROR;
 
     // Input metadata
     StreamMeta metadata{};
@@ -408,6 +418,8 @@ int Soapy_limesuiteng::readStreamStatus(
     SoapySDR::Stream* stream, [[maybe_unused]] size_t& chanMask, int& flags, long long& timeNs, const long timeoutUs)
 {
     auto icstream = reinterpret_cast<IConnectionStream*>(stream);
+    if (!rfstream)
+        return SOAPY_SDR_STREAM_ERROR;
     int ret = 0;
     flags = 0;
     StreamStats metadata;

